File: src/main/java/org/luxons/sevenwonders/repositories/LobbyRepository.java
package org.luxons.sevenwonders.repositories;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import org.luxons.sevenwonders.game.Lobby;
import org.luxons.sevenwonders.game.Player;
import org.luxons.sevenwonders.game.data.GameDefinitionLoader;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Repository;

@Repository
public class LobbyRepository {

    private final GameDefinitionLoader gameDefinitionLoader;

    private Map<Long, Lobby> lobbies = new HashMap<>();

    private long lastGameId = 0;

    @Autowired
    public LobbyRepository(GameDefinitionLoader gameDefinitionLoader) {
        this.gameDefinitionLoader = gameDefinitionLoader;
    }

    public Collection<Lobby> list() {
        return lobbies.values();
    }

    public Lobby create(String gameName, Player owner) {
        long id = lastGameId++;
        Lobby lobby = new Lobby(id, gameName, owner, gameDefinitionLoader.getGameDefinition());
        lobbies.put(id, lobby);
        return lobby;
    }

    public Lobby find(long lobbyId) {
        Lobby lobby = lobbies.get(lobbyId);
        if (lobby == null) {
            throw new LobbyNotFoundException(String.valueOf(lobbyId));
        }
        return lobby;
    }

    public static class LobbyNotFoundException extends RuntimeException {
        LobbyNotFoundException(String name) {
            super("Lobby not found for game '" + name + "'");
        }
    }
}

File: sw-server/src/main/kotlin/org/luxons/sevenwonders/server/repositories/LobbyRepository.kt
package org.luxons.sevenwonders.server.repositories

import org.luxons.sevenwonders.engine.data.GameDefinition
import org.luxons.sevenwonders.server.lobby.Lobby
import org.luxons.sevenwonders.server.lobby.Player
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Repository
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong

@Repository
class LobbyRepository {

    private val lobbies = ConcurrentHashMap<Long, Lobby>()

    private var lastGameId: AtomicLong = AtomicLong(0)

    fun list(): Collection<Lobby> = lobbies.values

    fun create(gameName: String, owner: Player): Lobby {
        val id = lastGameId.getAndIncrement()
        val lobby = Lobby(id, gameName, owner, GameDefinition.load())
        lobbies[id] = lobby
        return lobby
    }

    fun find(lobbyId: Long): Lobby = lobbies[lobbyId] ?: throw LobbyNotFoundException(lobbyId)

    fun remove(lobbyId: Long): Lobby = lobbies.remove(lobbyId) ?: throw LobbyNotFoundException(lobbyId)
}

internal class LobbyNotFoundException(id: Long) : RuntimeException("Lobby not found for id '$id'")
